### PR TITLE
Parking Lot Routes

### PR DESCRIPTION
**Pull request by leogoldberg**
Added following routes:
get all parking lots
delete parking lot
put parking lot (update)

added caching for all parking lot data

fixed bug in create-parking, it was saving before qr code url was created, moved the save into the callback function of qrcodeurl and removed extraneous try/catch block

### CODE REVIEW

File: routes/parking.js [hunk 1]
```diff
@@ -1,7 +1,59 @@
...
+
+onInit();
+
+function onInit() {
+  createCache();
+}
+
+function createCache() {
+  Parking.find({}, async (err, parkingLots) => {
+
```
Comment (JasonXD-CS) [inline routes/parking.js:23]: why do we need caching when we have mongodb

File: routes/parking.js [hunk 2]
```diff
@@ -14,26 +66,68 @@ router.post('/create-parking', async (req, res) => {
...
+      .then(parking => {
+        addToCache(parking.toObject())
+        res.sendStatus(200)
+      })
+      .catch(err => res.status(400).send(`create parking failed ${err}`))
+  })
+})
+
+router.get('/all', async (req, res) => {
+  res.send(cachedParkingLots);
```
Comment (JasonXD-CS) [inline routes/parking.js:95]: you can just do `const parkingLots = Parking.find({ createdBy: userId})`

File: routes/parking.js [hunk 3]
```diff
@@ -14,26 +66,68 @@ router.post('/create-parking', async (req, res) => {
...
-      // If qrcode generated successfully, we save the document for the space.
-      // console.log(url);
-      parking.qrCodeUrl = url;
-    })
+  const parking = new Parking({
+    name: name,
+    number: number,
+    rate: rate,
+    address: address,
+
```
Comment (JasonXD-CS) [inline routes/parking.js]: add userId to `createdBy`

File: routes/parking.js [hunk 4]
```diff
@@ -14,26 +66,68 @@ router.post('/create-parking', async (req, res) => {
...
+  const { name, number, rate, address, qrCodeUrl } = req.body
+
+  if (!name || !number || !rate || !address || !qrCodeUrl) {
+    return res.status(400).send('Missing parameter')
   }
+
+  const parkingLot = { name, number, rate, address, qrCodeUrl }
+
+  console.log(parkingLot)
+  Parking.findOneAndReplace({ "_id": ObjectId(req.params.parkingId) }, parkingLot, function (err, parkingLot) {
```
Comment (JasonXD-CS) [inline routes/parking.js]: Use `findOneAndUpdate` instead of `findOneAndReplace`

File: routes/parking.js [hunk 5]
```diff
@@ -14,26 +66,68 @@ router.post('/create-parking', async (req, res) => {
...
+// })
+
+router.delete('/:parkingId', async (req, res) => {
+  Parking.findByIdAndDelete(req.params.parkingId).exec()
+    .then(parking => deleteFromCache(req.params.parkingId))
+    .catch(err => res.status(400).send(`delete parking failed ${err}`))
+})
+
+router.put('/:parkingId', async (req, res) => {
+  const { name, number, rate, address, qrCodeUrl } = req.body
```
Comment (JasonXD-CS) [inline routes/parking.js]: You don't need qrCodeUrl for this.

File: routes/parking.js (hunk 4 above)
Comment (JasonXD-CS) [reply routes/parking.js]: or FindByIdAndUpdate would be easier

File: routes/parking.js (hunk 1 above)
Comment (leogoldberg) [reply routes/parking.js:23]: All calls to mongo db are async, data does not get pulled right away. It's better for responsiveness to cache the data in memory imo. Also, we're only storing session id's in the parking lot, not the session object, and fetching the session objects for each parking lot every time the parking lot data is fetched is inefficient, so I think it' better to just fetch it once and then update the cache

File: routes/parking.js (hunk 2 above)
Comment (leogoldberg) [reply routes/parking.js:95]: Aren't all parking lots created by the same, hard-coded admin user id now?

File: routes/parking.js (hunk 4 above)
Comment (leogoldberg) [reply routes/parking.js]: We could do this, but I wanted to keep it as general as possible, as we're not sure which field in the parking lot the admin will choose to update (could be just the name, could be the name and address, etc) so i was thinking to have the react app just send the entire parking lot object over.

**Review by JasonXD-CS — APPROVED** [review]

**Pull request merged**